hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java

```diff
@@ -1,81 +1,122 @@
-            BasicAWSCredentials awsCreds = new BasicAWSCredentials(configConfiguration.getSelfService().getS3AccessKeyId(),
-                    configConfiguration.getSelfService().getS3SecretKeyId());
-            s3Client = AmazonS3ClientBuilder.standard()
-                    .withRegion(configConfiguration.getSelfService().getAwsRegion())
-                    .withCredentials(new AWSStaticCredentialsProvider(awsCreds))
-                    .build();
-        } else {
-            s3Client = AmazonS3ClientBuilder.defaultClient();
+    public RemoteConfigCollection getRemoteConfig(){
+        if ( selfServiceConfig == null || !selfServiceConfig.isEnabled()){
+            return EMPTY_COLLECTION;
+        }
+        try {
+            RemoteConfigCollection config = cache.get(selfServiceConfig.getS3ObjectKey());
+            return config;
+        } catch (ExecutionException e) {
+            LOG.warn("Unable to get {} from cache.", selfServiceConfig.getS3ObjectKey());
+            return EMPTY_COLLECTION;
         }
-        return new S3ConfigSource(configConfiguration, s3Client);
     }
 
-    public RemoteConfigCollection getRemoteConfig() {
-        S3Object fullObject = s3Client.getObject(
-                new GetObjectRequest(configConfiguration.getSelfService().getS3BucketName(),
-                configConfiguration.getSelfService().getS3ObjectKey()));
-        InputStream s3ObjectStream =  fullObject.getObjectContent();
-        ObjectMapper om = new ObjectMapper();
-        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
-        om.setDateFormat(df);
-        SimpleModule module = new SimpleModule();
-        module.addDeserializer(RemoteConfigCollection.class, new RemoteConfigCollectionDeserializer());
-        om.registerModule(module);
+    private RemoteConfigCollection mapToRemoteConfigCollection(InputStream inputStream, Date lastModified) throws IOException {
         try {
-            return om.readValue(s3ObjectStream, RemoteConfigCollection.class);
-        } catch(IOException e) {
-            LOG.error("An error occured trying to get or process object {} from S3 Bucket {}",
-                    configConfiguration.getSelfService().getS3ObjectKey(),
-                    configConfiguration.getSelfService().getS3BucketName(), e);
-            throw new RuntimeException(e);
+            SelfServiceMetadata metadata = objectMapper.readValue(inputStream, SelfServiceMetadata.class);
+            RemoteConfigCollection remoteConfig = new RemoteConfigCollection(lastModified, metadata);
+            return remoteConfig;
         } finally {
-            if(fullObject != null) {
-                try {
-                    fullObject.close();
-                } catch(IOException e) {
-                    throw new RuntimeException(e);
-                }
-            }
+            inputStream.close();
         }
     }
 
+    private S3Object getS3Object(String key) {
```

Comment:
There's alot of S3 client and object code scattered about. You could create a class that wraps the S3Client and the Objectmapper that behaves similar to our http clients. I think it would make our code a lot simpler if we could do something like:

```
client.get(SelfServiceMetadata.class)
```

Reply:
Can we have a think about that after this PR?

Reply:
sure